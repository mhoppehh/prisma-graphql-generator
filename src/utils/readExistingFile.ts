import { promises as fs } from 'fs'

export async function readExistingFile(path: string): Promise<string | null> {
  try {
    return await fs.readFile(path, 'utf8')
  } catch {
    return null
  }
}
