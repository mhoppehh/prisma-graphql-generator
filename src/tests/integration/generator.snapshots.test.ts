// Mock ts-morph before any imports that might use it
jest.mock('ts-morph', () => ({
  Project: jest.fn().mockImplementation(() => ({
    addSourceFileAtPath: jest.fn(),
    getSourceFiles: jest.fn(() => []),
    saveSync: jest.fn(),
  })),
  SyntaxKind: {
    MethodDeclaration: 'MethodDeclaration',
    PropertySignature: 'PropertySignature',
  },
}))

import { onGenerate } from '../../generator'
import { fileExists } from '../../utils/fileExists'
import { readExistingFile } from '../../utils/readExistingFile'
import { writeFileSafely } from '../../utils/writeFileSafely'
import { generateParameterCombinations } from '../../utils/testCombinatorics'
import path from 'path'
import options from '../__fixtures__/generated/options.patched'

jest.mock('prettier', () => ({
  format: (content: string) => content,
  resolveConfig: () => Promise.resolve(undefined),
}))

jest.setTimeout(15000)

jest.mock('fs', () => ({
  promises: {
    writeFile: jest.fn().mockResolvedValue(undefined),
    readFile: jest.fn().mockResolvedValue('enum SortOrder { asc desc }'),
  },
}))

jest.mock('@prisma/internals', () => ({
  logger: {
    info: jest.fn(),
    error: jest.fn(),
    warn: jest.fn(),
    debug: jest.fn(),
  },
}))

jest.mock('../../utils/writeFileSafely')
jest.mock('../../utils/readExistingFile')
jest.mock('../../utils/fileExists')

const mockWriteFileSafely = writeFileSafely as jest.MockedFunction<
  typeof writeFileSafely
>
const mockReadExistingFile = readExistingFile as jest.MockedFunction<
  typeof readExistingFile
>
const mockFileExists = fileExists as jest.MockedFunction<typeof fileExists>

const captureGeneratedOutput = () => {
// Helper function to capture and normalize generated files for snapshots
  const calls = mockWriteFileSafely.mock.calls

  
  const filteredCalls = calls.filter(call => {
      const filePath = call[0]
      const fileName = path.basename(filePath)
      const extension = path.extname(fileName)
      
      // Exclude all JSON files (especially options.json with DMMF data)
      if (extension === '.json') {
        return false
      }
      
      // Exclude schema.ts (contains generated TypeScript types)
      if (fileName === 'schema.ts') {
        return false
      }
      
      // Only include actual generated GraphQL and resolver files
      const shouldInclude = (filePath.includes('.graphql') || 
              filePath.includes('.resolver.') || 
              filePath.includes('.resolvers.')) &&
             !filePath.includes('dmmf') &&
             !filePath.includes('config')
      
      return shouldInclude
    })
  
  return filteredCalls
    .map(call => ({
      filePath: call[0]
        .replace(process.cwd(), '<PROJECT_ROOT>')
        .replace(/\\/g, '/'), // Normalize path separators
      content: call[1]
        .replace(/\/\* Generated at .+ \*\//, '/* Generated at <TIMESTAMP> */')
        .replace(/Generated on: .+/, 'Generated on: <DATE>')
        .replace(/\r\n/g, '\n') // Normalize line endings
        .trim() // Remove leading/trailing whitespace
    }))
    .filter(file => file.content.length > 0) // Remove empty files
}

describe('Generator Cartesian Product Snapshot Tests', () => {
  beforeEach(() => {
    jest.clearAllMocks()
    Object.keys(process.env)
      .filter(key => key.startsWith('GENERATOR_'))
      .forEach(key => delete process.env[key])
    mockFileExists.mockImplementation(() => Promise.resolve(false))
    mockReadExistingFile.mockImplementation(() => Promise.resolve(''))
    mockWriteFileSafely.mockImplementation(() => Promise.resolve(undefined))
  })

  describe('Core Generator Output Snapshots', () => {
    // Define key combinations we want to snapshot
    const snapshotParameters = {
      scenarios: {
        values: [
          { model: 'Employee', queries: ['findMany'], mutations: [], description: 'findMany-only' },
          { model: 'Employee', queries: ['findUnique'], mutations: [], description: 'findUnique-only' },
          { model: 'Employee', queries: [], mutations: ['create'], description: 'create-only' },
          { model: 'Employee', queries: ['findMany', 'findUnique'], mutations: ['create', 'update'], description: 'mixed-operations' },
          { model: 'Category', queries: ['findMany'], mutations: [], description: 'different-model' },
        ],
        multi: false
      }
    }

    const combinations = generateParameterCombinations(snapshotParameters)

    describe.each(combinations)(
      'Generator output scenarios',
      (scenarios) => {
        const scenario = scenarios[0]

        it(`should generate correct output for ${scenario.model} with ${scenario.description}`, async () => {
          // Set environment variables based on scenario
          process.env.GENERATOR_MODEL = scenario.model
          process.env.GENERATOR_MODULE_PATH = './src/modules'
          
          if (scenario.queries.length > 0) {
            process.env.GENERATOR_QUERIES = scenario.queries.join(',')
          }
          if (scenario.mutations.length > 0) {
            process.env.GENERATOR_MUTATIONS = scenario.mutations.join(',')
          }

          // Run the generator
          await onGenerate(options)

          // Capture and snapshot the output
          const generatedOutput = captureGeneratedOutput()
          
          const snapshot = {
            scenario: {
              model: scenario.model,
              queries: scenario.queries,
              mutations: scenario.mutations,
              description: scenario.description
            },
            files: generatedOutput
          }

          // Create snapshot with descriptive name
          expect(snapshot).toMatchSnapshot(`generator-output-${scenario.model}-${scenario.description}`)
        })
      }
    )
  })

  describe('File Merging Scenarios', () => {
    const mergingParameters = {
      scenarios: {
        values: [
          { 
            model: 'Employee', 
            queries: ['findMany'], 
            mutations: [], 
            existingType: 'sdl',
            description: 'merge-with-existing-sdl' 
          },
          { 
            model: 'Employee', 
            queries: [], 
            mutations: ['create'], 
            existingType: 'none',
            description: 'fresh-mutation-only' 
          }
        ],
        multi: false
      }
    }

    const mergingCombinations = generateParameterCombinations(mergingParameters)

    describe.each(mergingCombinations)(
      'File merging scenarios',
      (scenarios) => {
        const scenario = scenarios[0]

        it(`should handle ${scenario.description} for ${scenario.model}`, async () => {
          // Setup file existence mocking
          if (scenario.existingType === 'sdl') {
            mockFileExists.mockImplementation((filePath: string) => 
              Promise.resolve(filePath.includes('.sdl'))
            )
            mockReadExistingFile.mockImplementation((filePath: string) => {
              if (filePath.includes('.sdl')) {
                return Promise.resolve(`
type ${scenario.model} {
  id: ID!
  # Existing field
  existingField: String
}

type Query {
  existingQuery: String
}`)
              }
              return Promise.resolve('')
            })
          }

          // Set environment variables
          process.env.GENERATOR_MODEL = scenario.model
          process.env.GENERATOR_MODULE_PATH = './src/modules'
          
          if (scenario.queries.length > 0) {
            process.env.GENERATOR_QUERIES = scenario.queries.join(',')
          }
          if (scenario.mutations.length > 0) {
            process.env.GENERATOR_MUTATIONS = scenario.mutations.join(',')
          }

          // Run the generator
          await onGenerate(options)

          // Capture and snapshot the output
          const generatedOutput = captureGeneratedOutput()
          
          const snapshot = {
            scenario: {
              model: scenario.model,
              queries: scenario.queries,
              mutations: scenario.mutations,
              existingType: scenario.existingType,
              description: scenario.description
            },
            files: generatedOutput
          }

          expect(snapshot).toMatchSnapshot(`file-merging-${scenario.model}-${scenario.description}`)
        })
      }
    )
  })

  describe('Multiple Query/Mutation Combinations', () => {
    const multiOpParameters = {
      combinations: {
        values: [
          { 
            model: 'Product', 
            queries: ['findMany', 'findUnique'], 
            mutations: ['create', 'update'],
            description: 'common-operations'
          },
          { 
            model: 'Category', 
            queries: ['findUnique'], 
            mutations: ['update'],
            description: 'minimal-operations'
          }
        ],
        multi: false
      }
    }

    const multiOpCombinations = generateParameterCombinations(multiOpParameters)

    describe.each(multiOpCombinations)(
      'Multiple operation scenarios',
      (combinations) => {
        const combo = combinations[0]

        it(`should generate ${combo.description} for ${combo.model}`, async () => {
          process.env.GENERATOR_MODEL = combo.model
          process.env.GENERATOR_MODULE_PATH = './src/modules'
          process.env.GENERATOR_QUERIES = combo.queries.join(',')
          process.env.GENERATOR_MUTATIONS = combo.mutations.join(',')

          await onGenerate(options)

          const generatedOutput = captureGeneratedOutput()
          
          const snapshot = {
            scenario: {
              model: combo.model,
              queries: combo.queries,
              mutations: combo.mutations,
              description: combo.description
            },
            files: generatedOutput
          }

          expect(snapshot).toMatchSnapshot(`multi-ops-${combo.model}-${combo.description}`)
        })
      }
    )
  })
})
