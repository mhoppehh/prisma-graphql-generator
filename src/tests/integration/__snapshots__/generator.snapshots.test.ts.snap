// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Generator Cartesian Product Snapshot Tests Core Generator Output Snapshots Generator output scenarios should generate correct output for Category with different-model: generator-output-Category-different-model 1`] = `
{
  "files": [
    {
      "content": "type Category {
  CategoryName: String
  Description: String
  Id: Int!
}

input CategoryWhereInput {
  AND: CategoryWhereInput
  OR: [CategoryWhereInput!]
  NOT: CategoryWhereInput
  CategoryName: String
  Description: String
  Id: Int
}

input CategoryOrderByWithRelationInput {
  CategoryName: SortOrder
  Description: SortOrder
  Id: SortOrder
}


type Category {
  CategoryName: String
  Description: String
  Id: Int
}


extend type Query {
  categoriesFindMany(where: CategoryWhereInput, orderBy: CategoryOrderByWithRelationInput, cursor: CategoryWhereUniqueInput, take: Int, skip: Int, distinct: CategoryScalarFieldEnum): [Category]
}",
      "filePath": "src/modules/Category.graphql",
    },
    {
      "content": "import { CategoryModule } from "./module-types"

export const resolversCategory: CategoryModule.Resolvers = {
  Query: {
    categoriesFindMany: async (_parent, args, context) => {
      return (await context.dataSources.prisma()).category.findMany({
        where: args.where,
        orderBy: args.orderBy,
        take: args.take,
        skip: args.skip,
      });
    },
  },

};",
      "filePath": "src/modules/Category.resolver.ts",
    },
  ],
  "scenario": {
    "description": "different-model",
    "model": "Category",
    "mutations": [],
    "queries": [
      "findMany",
    ],
  },
}
`;

exports[`Generator Cartesian Product Snapshot Tests Core Generator Output Snapshots Generator output scenarios should generate correct output for Employee with create-only: generator-output-Employee-create-only 1`] = `
{
  "files": [
    {
      "content": "type Employee {
  Address: String
  BirthDate: String
  City: String
  Country: String
  Extension: String
  FirstName: String
  HireDate: String
  HomePhone: String
  Id: Int!
  LastName: String
  Notes: String
  Photo: String
  PhotoPath: String
  PostalCode: String
  Region: String
  ReportsTo: Int
  Title: String
  TitleOfCourtesy: String
}

input EmployeeCreateInput {
  Address: String
  BirthDate: String
  City: String
  Country: String
  Extension: String
  FirstName: String
  HireDate: String
  HomePhone: String
  Id: Int!
  LastName: String
  Notes: String
  Photo: String
  PhotoPath: String
  PostalCode: String
  Region: String
  ReportsTo: Int
  Title: String
  TitleOfCourtesy: String
}


type Employee {
  Address: String
  BirthDate: String
  City: String
  Country: String
  Extension: String
  FirstName: String
  HireDate: String
  HomePhone: String
  Id: Int
  LastName: String
  Notes: String
  Photo: String
  PhotoPath: String
  PostalCode: String
  Region: String
  ReportsTo: Int
  Title: String
  TitleOfCourtesy: String
}



extend type Mutation {
  employeeCreate(data: EmployeeCreateInput!): Employee
}",
      "filePath": "src/modules/Employee.graphql",
    },
    {
      "content": "import { EmployeeModule } from "./module-types"

export const resolversEmployee: EmployeeModule.Resolvers = {

  Mutation: {
    employeeCreate: async (_parent, args, context) => {
      return (await context.dataSources.prisma()).employee.create({
        data: args.data,
      });
    },
  },
};",
      "filePath": "src/modules/Employee.resolver.ts",
    },
  ],
  "scenario": {
    "description": "create-only",
    "model": "Employee",
    "mutations": [
      "create",
    ],
    "queries": [],
  },
}
`;

exports[`Generator Cartesian Product Snapshot Tests Core Generator Output Snapshots Generator output scenarios should generate correct output for Employee with findMany-only: generator-output-Employee-findMany-only 1`] = `
{
  "files": [
    {
      "content": "type Employee {
  Address: String
  BirthDate: String
  City: String
  Country: String
  Extension: String
  FirstName: String
  HireDate: String
  HomePhone: String
  Id: Int!
  LastName: String
  Notes: String
  Photo: String
  PhotoPath: String
  PostalCode: String
  Region: String
  ReportsTo: Int
  Title: String
  TitleOfCourtesy: String
}

input EmployeeWhereInput {
  AND: EmployeeWhereInput
  OR: [EmployeeWhereInput!]
  NOT: EmployeeWhereInput
  Address: String
  BirthDate: String
  City: String
  Country: String
  Extension: String
  FirstName: String
  HireDate: String
  HomePhone: String
  Id: Int
  LastName: String
  Notes: String
  Photo: String
  PhotoPath: String
  PostalCode: String
  Region: String
  ReportsTo: Int
  Title: String
  TitleOfCourtesy: String
}

input EmployeeOrderByWithRelationInput {
  Address: SortOrder
  BirthDate: SortOrder
  City: SortOrder
  Country: SortOrder
  Extension: SortOrder
  FirstName: SortOrder
  HireDate: SortOrder
  HomePhone: SortOrder
  Id: SortOrder
  LastName: SortOrder
  Notes: SortOrder
  Photo: SortOrder
  PhotoPath: SortOrder
  PostalCode: SortOrder
  Region: SortOrder
  ReportsTo: SortOrder
  Title: SortOrder
  TitleOfCourtesy: SortOrder
}


type Employee {
  Address: String
  BirthDate: String
  City: String
  Country: String
  Extension: String
  FirstName: String
  HireDate: String
  HomePhone: String
  Id: Int
  LastName: String
  Notes: String
  Photo: String
  PhotoPath: String
  PostalCode: String
  Region: String
  ReportsTo: Int
  Title: String
  TitleOfCourtesy: String
}


extend type Query {
  employeesFindMany(where: EmployeeWhereInput, orderBy: EmployeeOrderByWithRelationInput, cursor: EmployeeWhereUniqueInput, take: Int, skip: Int, distinct: EmployeeScalarFieldEnum): [Employee]
}",
      "filePath": "src/modules/Employee.graphql",
    },
    {
      "content": "import { EmployeeModule } from "./module-types"

export const resolversEmployee: EmployeeModule.Resolvers = {
  Query: {
    employeesFindMany: async (_parent, args, context) => {
      return (await context.dataSources.prisma()).employee.findMany({
        where: args.where,
        orderBy: args.orderBy,
        take: args.take,
        skip: args.skip,
      });
    },
  },

};",
      "filePath": "src/modules/Employee.resolver.ts",
    },
  ],
  "scenario": {
    "description": "findMany-only",
    "model": "Employee",
    "mutations": [],
    "queries": [
      "findMany",
    ],
  },
}
`;

exports[`Generator Cartesian Product Snapshot Tests Core Generator Output Snapshots Generator output scenarios should generate correct output for Employee with findUnique-only: generator-output-Employee-findUnique-only 1`] = `
{
  "files": [
    {
      "content": "type Employee {
  Address: String
  BirthDate: String
  City: String
  Country: String
  Extension: String
  FirstName: String
  HireDate: String
  HomePhone: String
  Id: Int!
  LastName: String
  Notes: String
  Photo: String
  PhotoPath: String
  PostalCode: String
  Region: String
  ReportsTo: Int
  Title: String
  TitleOfCourtesy: String
}

input EmployeeWhereInput {
  AND: EmployeeWhereInput
  OR: [EmployeeWhereInput!]
  NOT: EmployeeWhereInput
  Address: String
  BirthDate: String
  City: String
  Country: String
  Extension: String
  FirstName: String
  HireDate: String
  HomePhone: String
  Id: Int
  LastName: String
  Notes: String
  Photo: String
  PhotoPath: String
  PostalCode: String
  Region: String
  ReportsTo: Int
  Title: String
  TitleOfCourtesy: String
}

input EmployeeWhereUniqueInput {
  Id: Int
  AND: EmployeeWhereInput
  OR: [EmployeeWhereInput!]
  NOT: EmployeeWhereInput
  Address: String
  BirthDate: String
  City: String
  Country: String
  Extension: String
  FirstName: String
  HireDate: String
  HomePhone: String
  LastName: String
  Notes: String
  Photo: String
  PhotoPath: String
  PostalCode: String
  Region: String
  ReportsTo: Int
  Title: String
  TitleOfCourtesy: String
}


type Employee {
  Address: String
  BirthDate: String
  City: String
  Country: String
  Extension: String
  FirstName: String
  HireDate: String
  HomePhone: String
  Id: Int
  LastName: String
  Notes: String
  Photo: String
  PhotoPath: String
  PostalCode: String
  Region: String
  ReportsTo: Int
  Title: String
  TitleOfCourtesy: String
}


extend type Query {
  employeeFindUnique(where: EmployeeWhereUniqueInput!): Employee!
}",
      "filePath": "src/modules/Employee.graphql",
    },
    {
      "content": "import { EmployeeModule } from "./module-types"

export const resolversEmployee: EmployeeModule.Resolvers = {
  Query: {
    employeeFindUnique: async (_parent, args, context) => {
      return (await context.dataSources.prisma()).employee.findUnique({
        where: args.where,
      });
    },
  },

};",
      "filePath": "src/modules/Employee.resolver.ts",
    },
  ],
  "scenario": {
    "description": "findUnique-only",
    "model": "Employee",
    "mutations": [],
    "queries": [
      "findUnique",
    ],
  },
}
`;

exports[`Generator Cartesian Product Snapshot Tests Core Generator Output Snapshots Generator output scenarios should generate correct output for Employee with mixed-operations: generator-output-Employee-mixed-operations 1`] = `
{
  "files": [
    {
      "content": "type Employee {
  Address: String
  BirthDate: String
  City: String
  Country: String
  Extension: String
  FirstName: String
  HireDate: String
  HomePhone: String
  Id: Int!
  LastName: String
  Notes: String
  Photo: String
  PhotoPath: String
  PostalCode: String
  Region: String
  ReportsTo: Int
  Title: String
  TitleOfCourtesy: String
}

input EmployeeWhereInput {
  AND: EmployeeWhereInput
  OR: [EmployeeWhereInput!]
  NOT: EmployeeWhereInput
  Address: String
  BirthDate: String
  City: String
  Country: String
  Extension: String
  FirstName: String
  HireDate: String
  HomePhone: String
  Id: Int
  LastName: String
  Notes: String
  Photo: String
  PhotoPath: String
  PostalCode: String
  Region: String
  ReportsTo: Int
  Title: String
  TitleOfCourtesy: String
}

input EmployeeOrderByWithRelationInput {
  Address: SortOrder
  BirthDate: SortOrder
  City: SortOrder
  Country: SortOrder
  Extension: SortOrder
  FirstName: SortOrder
  HireDate: SortOrder
  HomePhone: SortOrder
  Id: SortOrder
  LastName: SortOrder
  Notes: SortOrder
  Photo: SortOrder
  PhotoPath: SortOrder
  PostalCode: SortOrder
  Region: SortOrder
  ReportsTo: SortOrder
  Title: SortOrder
  TitleOfCourtesy: SortOrder
}

input EmployeeWhereUniqueInput {
  Id: Int
  AND: EmployeeWhereInput
  OR: [EmployeeWhereInput!]
  NOT: EmployeeWhereInput
  Address: String
  BirthDate: String
  City: String
  Country: String
  Extension: String
  FirstName: String
  HireDate: String
  HomePhone: String
  LastName: String
  Notes: String
  Photo: String
  PhotoPath: String
  PostalCode: String
  Region: String
  ReportsTo: Int
  Title: String
  TitleOfCourtesy: String
}

input EmployeeCreateInput {
  Address: String
  BirthDate: String
  City: String
  Country: String
  Extension: String
  FirstName: String
  HireDate: String
  HomePhone: String
  Id: Int!
  LastName: String
  Notes: String
  Photo: String
  PhotoPath: String
  PostalCode: String
  Region: String
  ReportsTo: Int
  Title: String
  TitleOfCourtesy: String
}

input EmployeeUpdateInput {
  Address: String
  BirthDate: String
  City: String
  Country: String
  Extension: String
  FirstName: String
  HireDate: String
  HomePhone: String
  Id: Int
  LastName: String
  Notes: String
  Photo: String
  PhotoPath: String
  PostalCode: String
  Region: String
  ReportsTo: Int
  Title: String
  TitleOfCourtesy: String
}


type Employee {
  Address: String
  BirthDate: String
  City: String
  Country: String
  Extension: String
  FirstName: String
  HireDate: String
  HomePhone: String
  Id: Int
  LastName: String
  Notes: String
  Photo: String
  PhotoPath: String
  PostalCode: String
  Region: String
  ReportsTo: Int
  Title: String
  TitleOfCourtesy: String
}


extend type Query {
  employeesFindMany(where: EmployeeWhereInput, orderBy: EmployeeOrderByWithRelationInput, cursor: EmployeeWhereUniqueInput, take: Int, skip: Int, distinct: EmployeeScalarFieldEnum): [Employee]
  employeeFindUnique(where: EmployeeWhereUniqueInput!): Employee!
}

extend type Mutation {
  employeeCreate(data: EmployeeCreateInput!): Employee
  employeeUpdate(data: EmployeeUpdateInput!, where: EmployeeWhereUniqueInput!): Employee!
}",
      "filePath": "src/modules/Employee.graphql",
    },
    {
      "content": "import { EmployeeModule } from "./module-types"

export const resolversEmployee: EmployeeModule.Resolvers = {
  Query: {
    employeesFindMany: async (_parent, args, context) => {
      return (await context.dataSources.prisma()).employee.findMany({
        where: args.where,
        orderBy: args.orderBy,
        take: args.take,
        skip: args.skip,
      });
    },
    employeeFindUnique: async (_parent, args, context) => {
      return (await context.dataSources.prisma()).employee.findUnique({
        where: args.where,
      });
    },
  },

  Mutation: {
    employeeCreate: async (_parent, args, context) => {
      return (await context.dataSources.prisma()).employee.create({
        data: args.data,
      });
    },
    employeeUpdate: async (_parent, args, context) => {
      return (await context.dataSources.prisma()).employee.update({
        where: args.where,
        data: args.data,
      });
    },
  },
};",
      "filePath": "src/modules/Employee.resolver.ts",
    },
  ],
  "scenario": {
    "description": "mixed-operations",
    "model": "Employee",
    "mutations": [
      "create",
      "update",
    ],
    "queries": [
      "findMany",
      "findUnique",
    ],
  },
}
`;

exports[`Generator Cartesian Product Snapshot Tests File Merging Scenarios File merging scenarios should handle fresh-mutation-only for Employee: file-merging-Employee-fresh-mutation-only 1`] = `
{
  "files": [
    {
      "content": "type Employee {
  Address: String
  BirthDate: String
  City: String
  Country: String
  Extension: String
  FirstName: String
  HireDate: String
  HomePhone: String
  Id: Int!
  LastName: String
  Notes: String
  Photo: String
  PhotoPath: String
  PostalCode: String
  Region: String
  ReportsTo: Int
  Title: String
  TitleOfCourtesy: String
}

input EmployeeCreateInput {
  Address: String
  BirthDate: String
  City: String
  Country: String
  Extension: String
  FirstName: String
  HireDate: String
  HomePhone: String
  Id: Int!
  LastName: String
  Notes: String
  Photo: String
  PhotoPath: String
  PostalCode: String
  Region: String
  ReportsTo: Int
  Title: String
  TitleOfCourtesy: String
}


type Employee {
  Address: String
  BirthDate: String
  City: String
  Country: String
  Extension: String
  FirstName: String
  HireDate: String
  HomePhone: String
  Id: Int
  LastName: String
  Notes: String
  Photo: String
  PhotoPath: String
  PostalCode: String
  Region: String
  ReportsTo: Int
  Title: String
  TitleOfCourtesy: String
}



extend type Mutation {
  employeeCreate(data: EmployeeCreateInput!): Employee
}",
      "filePath": "src/modules/Employee.graphql",
    },
    {
      "content": "import { EmployeeModule } from "./module-types"

export const resolversEmployee: EmployeeModule.Resolvers = {

  Mutation: {
    employeeCreate: async (_parent, args, context) => {
      return (await context.dataSources.prisma()).employee.create({
        data: args.data,
      });
    },
  },
};",
      "filePath": "src/modules/Employee.resolver.ts",
    },
  ],
  "scenario": {
    "description": "fresh-mutation-only",
    "existingType": "none",
    "model": "Employee",
    "mutations": [
      "create",
    ],
    "queries": [],
  },
}
`;

exports[`Generator Cartesian Product Snapshot Tests File Merging Scenarios File merging scenarios should handle merge-with-existing-sdl for Employee: file-merging-Employee-merge-with-existing-sdl 1`] = `
{
  "files": [
    {
      "content": "type Employee {
  Address: String
  BirthDate: String
  City: String
  Country: String
  Extension: String
  FirstName: String
  HireDate: String
  HomePhone: String
  Id: Int!
  LastName: String
  Notes: String
  Photo: String
  PhotoPath: String
  PostalCode: String
  Region: String
  ReportsTo: Int
  Title: String
  TitleOfCourtesy: String
}

input EmployeeWhereInput {
  AND: EmployeeWhereInput
  OR: [EmployeeWhereInput!]
  NOT: EmployeeWhereInput
  Address: String
  BirthDate: String
  City: String
  Country: String
  Extension: String
  FirstName: String
  HireDate: String
  HomePhone: String
  Id: Int
  LastName: String
  Notes: String
  Photo: String
  PhotoPath: String
  PostalCode: String
  Region: String
  ReportsTo: Int
  Title: String
  TitleOfCourtesy: String
}

input EmployeeOrderByWithRelationInput {
  Address: SortOrder
  BirthDate: SortOrder
  City: SortOrder
  Country: SortOrder
  Extension: SortOrder
  FirstName: SortOrder
  HireDate: SortOrder
  HomePhone: SortOrder
  Id: SortOrder
  LastName: SortOrder
  Notes: SortOrder
  Photo: SortOrder
  PhotoPath: SortOrder
  PostalCode: SortOrder
  Region: SortOrder
  ReportsTo: SortOrder
  Title: SortOrder
  TitleOfCourtesy: SortOrder
}


type Employee {
  Address: String
  BirthDate: String
  City: String
  Country: String
  Extension: String
  FirstName: String
  HireDate: String
  HomePhone: String
  Id: Int
  LastName: String
  Notes: String
  Photo: String
  PhotoPath: String
  PostalCode: String
  Region: String
  ReportsTo: Int
  Title: String
  TitleOfCourtesy: String
}


extend type Query {
  employeesFindMany(where: EmployeeWhereInput, orderBy: EmployeeOrderByWithRelationInput, cursor: EmployeeWhereUniqueInput, take: Int, skip: Int, distinct: EmployeeScalarFieldEnum): [Employee]
}",
      "filePath": "src/modules/Employee.graphql",
    },
    {
      "content": "import { EmployeeModule } from "./module-types"

export const resolversEmployee: EmployeeModule.Resolvers = {
  Query: {
    employeesFindMany: async (_parent, args, context) => {
      return (await context.dataSources.prisma()).employee.findMany({
        where: args.where,
        orderBy: args.orderBy,
        take: args.take,
        skip: args.skip,
      });
    },
  },

};",
      "filePath": "src/modules/Employee.resolver.ts",
    },
  ],
  "scenario": {
    "description": "merge-with-existing-sdl",
    "existingType": "sdl",
    "model": "Employee",
    "mutations": [],
    "queries": [
      "findMany",
    ],
  },
}
`;

exports[`Generator Cartesian Product Snapshot Tests Multiple Query/Mutation Combinations Multiple operation scenarios should generate common-operations for Product: multi-ops-Product-common-operations 1`] = `
{
  "files": [
    {
      "content": "type Product {
  CategoryId: Int!
  Discontinued: Int!
  Id: Int!
  ProductName: String
  QuantityPerUnit: String
  ReorderLevel: Int!
  SupplierId: Int!
  UnitPrice: String!
  UnitsInStock: Int!
  UnitsOnOrder: Int!
}

input ProductWhereInput {
  AND: ProductWhereInput
  OR: [ProductWhereInput!]
  NOT: ProductWhereInput
  CategoryId: Int
  Discontinued: Int
  Id: Int
  ProductName: String
  QuantityPerUnit: String
  ReorderLevel: Int
  SupplierId: Int
  UnitPrice: String
  UnitsInStock: Int
  UnitsOnOrder: Int
}

input ProductOrderByWithRelationInput {
  CategoryId: SortOrder
  Discontinued: SortOrder
  Id: SortOrder
  ProductName: SortOrder
  QuantityPerUnit: SortOrder
  ReorderLevel: SortOrder
  SupplierId: SortOrder
  UnitPrice: SortOrder
  UnitsInStock: SortOrder
  UnitsOnOrder: SortOrder
}

input ProductWhereUniqueInput {
  Id: Int
  AND: ProductWhereInput
  OR: [ProductWhereInput!]
  NOT: ProductWhereInput
  CategoryId: Int
  Discontinued: Int
  ProductName: String
  QuantityPerUnit: String
  ReorderLevel: Int
  SupplierId: Int
  UnitPrice: String
  UnitsInStock: Int
  UnitsOnOrder: Int
}

input ProductCreateInput {
  CategoryId: Int!
  Discontinued: Int!
  Id: Int!
  ProductName: String
  QuantityPerUnit: String
  ReorderLevel: Int!
  SupplierId: Int!
  UnitPrice: String!
  UnitsInStock: Int!
  UnitsOnOrder: Int!
}

input ProductUpdateInput {
  CategoryId: Int
  Discontinued: Int
  Id: Int
  ProductName: String
  QuantityPerUnit: String
  ReorderLevel: Int
  SupplierId: Int
  UnitPrice: String
  UnitsInStock: Int
  UnitsOnOrder: Int
}


type Product {
  CategoryId: Int
  Discontinued: Int
  Id: Int
  ProductName: String
  QuantityPerUnit: String
  ReorderLevel: Int
  SupplierId: Int
  UnitPrice: String
  UnitsInStock: Int
  UnitsOnOrder: Int
}


extend type Query {
  productsFindMany(where: ProductWhereInput, orderBy: ProductOrderByWithRelationInput, cursor: ProductWhereUniqueInput, take: Int, skip: Int, distinct: ProductScalarFieldEnum): [Product]
  productFindUnique(where: ProductWhereUniqueInput!): Product!
}

extend type Mutation {
  productCreate(data: ProductCreateInput!): Product
  productUpdate(data: ProductUpdateInput!, where: ProductWhereUniqueInput!): Product!
}",
      "filePath": "src/modules/Product.graphql",
    },
    {
      "content": "import { ProductModule } from "./module-types"

export const resolversProduct: ProductModule.Resolvers = {
  Query: {
    productsFindMany: async (_parent, args, context) => {
      return (await context.dataSources.prisma()).product.findMany({
        where: args.where,
        orderBy: args.orderBy,
        take: args.take,
        skip: args.skip,
      });
    },
    productFindUnique: async (_parent, args, context) => {
      return (await context.dataSources.prisma()).product.findUnique({
        where: args.where,
      });
    },
  },

  Mutation: {
    productCreate: async (_parent, args, context) => {
      return (await context.dataSources.prisma()).product.create({
        data: args.data,
      });
    },
    productUpdate: async (_parent, args, context) => {
      return (await context.dataSources.prisma()).product.update({
        where: args.where,
        data: args.data,
      });
    },
  },
};",
      "filePath": "src/modules/Product.resolver.ts",
    },
  ],
  "scenario": {
    "description": "common-operations",
    "model": "Product",
    "mutations": [
      "create",
      "update",
    ],
    "queries": [
      "findMany",
      "findUnique",
    ],
  },
}
`;

exports[`Generator Cartesian Product Snapshot Tests Multiple Query/Mutation Combinations Multiple operation scenarios should generate minimal-operations for Category: multi-ops-Category-minimal-operations 1`] = `
{
  "files": [
    {
      "content": "type Category {
  CategoryName: String
  Description: String
  Id: Int!
}

input CategoryWhereInput {
  AND: CategoryWhereInput
  OR: [CategoryWhereInput!]
  NOT: CategoryWhereInput
  CategoryName: String
  Description: String
  Id: Int
}

input CategoryWhereUniqueInput {
  Id: Int
  AND: CategoryWhereInput
  OR: [CategoryWhereInput!]
  NOT: CategoryWhereInput
  CategoryName: String
  Description: String
}

input CategoryUpdateInput {
  CategoryName: String
  Description: String
  Id: Int
}


type Category {
  CategoryName: String
  Description: String
  Id: Int
}


extend type Query {
  categoryFindUnique(where: CategoryWhereUniqueInput!): Category!
}

extend type Mutation {
  categoryUpdate(data: CategoryUpdateInput!, where: CategoryWhereUniqueInput!): Category!
}",
      "filePath": "src/modules/Category.graphql",
    },
    {
      "content": "import { CategoryModule } from "./module-types"

export const resolversCategory: CategoryModule.Resolvers = {
  Query: {
    categoryFindUnique: async (_parent, args, context) => {
      return (await context.dataSources.prisma()).category.findUnique({
        where: args.where,
      });
    },
  },

  Mutation: {
    categoryUpdate: async (_parent, args, context) => {
      return (await context.dataSources.prisma()).category.update({
        where: args.where,
        data: args.data,
      });
    },
  },
};",
      "filePath": "src/modules/Category.resolver.ts",
    },
  ],
  "scenario": {
    "description": "minimal-operations",
    "model": "Category",
    "mutations": [
      "update",
    ],
    "queries": [
      "findUnique",
    ],
  },
}
`;
